### PR TITLE
Fix bug when image resized

### LINKED ISSUE #18: Face bounding box is for scaled image
Before face detection we rescale the image:
https://github.com/IBM/MAX-Facial-Age-Estimator/blob/master/core/backend.py#L62

So the bounding box coordinates returned are for the scaled image and so not usable on the original image (e.g. to draw the bounding boxes).

e.g.

image

image

### PR DESCRIPTION
**Pull request by I-Hong**
This PR fixed the output of coordinates of bounding boxes when an image ran resize process. 
If the width or height of an input image is larger than 1024, we do resize the image for processing, but return the `scaled bounding boxes`. 

Fixes #18

### CODE REVIEW

**Comment by I-Hong** [comment]
Hi @MLnick, thank you for the comments. I've fixed the bug you mentioned https://github.com/IBM/MAX-Facial-Age-Estimator/issues/18#issue-418278269. I didn't add a flag in the code; instead, I check the size of input image, if one of the image size is larger than 1024, the input image will be resized.

File: core/model.py [hunk 1]
```diff
@@ -60,17 +72,20 @@ def __init__(self, path=DEFAULT_MODEL_PATH):
...
     def _pre_process(self, input_img):
         ad = 0.4
+        ratio=1
         img_h, img_w, _ = np.shape(input_img)
-        input_img = cv2.resize(input_img, (1024, int(1024 * img_h / img_w)))
-        img_h, img_w, _ = np.shape(input_img)
 
+        # if image size > 1024 then resize
+        if img_h > 1024 or img_w > 1024:
+            input_img, ratio =img_resize(input_img)
```
Comment (MLnick) [inline core/model.py]: We can just do `input_img, ratio = img_resize(input_img)` here, since the function returns the `input_data` unscaled if h/w <= 1024. 

Just add the line `ratio = 1` at the start of the `img_resize` function

File: core/model.py [hunk 2]
```diff
@@ -32,12 +32,24 @@ def read_still_image(image_data):
         logger.error(e)
         abort(400, 'Invalid file type/extension. Please provide a valid image (supported formats: JPEG, PNG, TIFF).')
 
+def img_resize(input_data):
+    img_h, img_w, _ = np.shape(input_data)
```
Comment (MLnick) [inline core/model.py]: Add in the line `ratio=1` from https://github.com/IBM/MAX-Facial-Age-Estimator/pull/19/files#diff-0831a56d210514c03372021d9e8f19fbR81

**Comment by MLnick** [comment]
@I-Hong some merge conflict needs to be resolved here.

I made a couple comments, overall seems good and I tested locally and the bbox coordinates work. However we will need some test cases for `img_resize`:

1. image smaller than 1024, check it comes back same size
2. image h > 1024, check scaled dimensions are as expected
3. image w > 1024, check scaled dimensions are as expected

File: core/model.py (hunk 1 above)
Comment (I-Hong) [reply core/model.py]: Thanks @MLnick, it's a neat code.

**Comment by I-Hong** [comment]
> 2\. mage h > 1024, check scaled dimensions are as expected

Thanks @MLnick. I've updated several resized images with different resolutions and a test file to see if the resized **back** images are consistent with input dimensions. But, I couldn't import `img_resize` function by `from core.model import img_resize`. I tested in my local side, it works. I'm not sure why it doesn't work at the repo test. Do you have any suggestions

File: core/model.py [hunk 3]
```diff
@@ -32,12 +32,24 @@ def read_still_image(image_data):
         logger.error(e)
         abort(400, 'Invalid file type/extension. Please provide a valid image (supported formats: JPEG, PNG, TIFF).')
 
+def img_resize(input_data, ratio):
```
Comment (MLnick) [inline core/model.py]: `ratio` here does not have an impact on how the method works and is in fact a returned value of the method. So it does not make sense for it to be an argument here. Put another way, why would you ever pass in `ratio` other than `1` to this method?

Just add the line `ratio = 1` as the first line in the method

File: tests/test.py [hunk 4]
```diff
@@ -57,11 +61,37 @@ def test_predict():
...
 
-    file_path = 'README.md'
+    file_path = '../README.md'
     with open(file_path,'rb') as file:
         file_form = {'text': (file_path, file, 'text/plain')}
         r = requests.post(url=model_endpoint, files=file_form)
     assert r.status_code == 400
 
+
+def _img_resize(input_data, ratio):
```
Comment (MLnick) [inline tests/test.py]: We need to test the actual method we use in the main code path - copying it here in the tests is not the best way to do this, so let's remove this one here.

File: tests/test.py [hunk 5]
```diff
@@ -30,16 +34,16 @@ def _check_response(r):
...
+    assert 335 > json['predictions'][0]['face_box'][1] > 320
+    assert 720 > json['predictions'][0]['face_box'][2] > 700
+    assert 975 > json['predictions'][0]['face_box'][3] > 960
 
 
 def test_predict():
     model_endpoint = 'http://localhost:5000/model/predict'
     formats = ['jpg', 'png', 'tiff']
-    file_path = 'tests/tom_cruise.{}'
+    file_path = 'tom_cruise.{}'
```
Comment (MLnick) [inline tests/test.py]: This is going to fail - we need to add back the `tests/` path - revert this change and the 2 others below.

File: tests/test.py [hunk 6]
```diff
@@ -48,7 +52,7 @@ def test_predict():
             r = requests.post(url=model_endpoint, files=file_form)
         _check_response(r)
 
-    file_path3 = 'tests/non_face.jpg'
+    file_path3 = 'non_face.jpg'
```
Comment (MLnick) [inline tests/test.py]: revert this (see above comment)

File: tests/test.py [hunk 7]
```diff
@@ -57,11 +61,37 @@ def test_predict():
     assert json['status'] == "ok"
     assert json['predictions'] ==[]
 
-    file_path = 'README.md'
+    file_path = '../README.md'
```
Comment (MLnick) [inline tests/test.py]: revert this (see above comment)

File: core/model.py (hunk 3 above)
Comment (I-Hong) [reply core/model.py]: Thanks @MLnick, I've updated it.

File: tests/test.py (hunk 5 above)
Comment (I-Hong) [reply tests/test.py]: Yes, it's different to my local test. I've updated it.

File: tests/test.py (hunk 4 above)
Comment (I-Hong) [reply tests/test.py]: Thanks @MLnick, I've moved it to `util.py`.

File: tests/test.py [hunk 8]
```diff
@@ -63,5 +67,24 @@ def test_predict():
...
 
+def test_img_resize():
+        """
+        The image resize test.
+        """
+
+        resize_path=[]
+        for file in os.listdir("."):
+            if file.startswith("resize_"):
+                resize_path.append(file)
```
Comment (MLnick) [inline tests/test.py]:
```suggestion
                resize_path.append("tests/" + file)
```

File: tests/test.py [hunk 9]
```diff
@@ -63,5 +67,24 @@ def test_predict():
         r = requests.post(url=model_endpoint, files=file_form)
     assert r.status_code == 400
 
+def test_img_resize():
+        """
+        The image resize test.
+        """
+
+        resize_path=[]
+        for file in os.listdir("."):
```
Comment (MLnick) [inline tests/test.py]: This test is not actually doing anything as `resize_path` will be an empty list `[]`. So none of the assertions in the `for` loop will be tested.

```suggestion
        for file in os.listdir("tests/"):
```

File: .travis.yml [hunk 10]
```diff
@@ -6,8 +6,9 @@ services:
 install:
   - docker build -t max-facial-age-estimator .
   - docker run -it -d -p 5000:5000 max-facial-age-estimator
+  - export PYTHONPATH=$PYTHONPATH:$(pwd)/core
```
Comment (MLnick) [inline .travis.yml]: According to [this link](https://docs.pytest.org/en/latest/pythonpath.html#invoking-pytest-versus-python-m-pytest), we can remove this line and below run `python -m pytest tests/test.py` - which will also add the current directory to sys path which solves the problem and I think is marginally cleaner.

File: .travis.yml [hunk 11]
```diff
@@ -6,8 +6,9 @@ services:
...
   - sleep 10
 before_script:
-  - pip install pytest requests
+  - pip install pytest requests opencv-python numpy Pillow
 script:
 - pytest tests/test.py
```
Comment (MLnick) [inline .travis.yml]:
```suggestion
- python -m pytest tests/test.py
```

File: tests/test.py (hunk 9 above)
Comment (I-Hong) [reply tests/test.py]: Thanks @MLnick.

File: .travis.yml (hunk 11 above)
Comment (I-Hong) [reply .travis.yml]: Thanks @MLnick, it works.
